### PR TITLE
http: split header decoding into concrete types

### PR DESCRIPTION
**Pull request by mattklein123**
This continues the refactor to used typed headers throughout
the code base. This change only implements the decoding path,
as encoding will require changing the HTTP filter interface.
This completes the minimum amount of changes that will allow
us to fix https://github.com/envoyproxy/envoy/issues/9873 in
a type safe and easy to understand way.

Risk Level: Medium (HTTP/1 codec changes, though good coverage)
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: source/common/http/http1/codec_impl.cc [hunk 1]
```diff
@@ -926,10 +922,12 @@ void ClientConnectionImpl::onMessageComplete(HeaderMapImplPtr&& trailers) {
     }
 
     if (deferred_end_stream_headers_) {
-      response.decoder_->decodeHeaders(std::move(deferred_end_stream_headers_), true);
-      deferred_end_stream_headers_.reset();
+      response.decoder_->decodeHeaders(
+          absl::get<ResponseHeaderMapImplPtr>(std::move(headers_or_trailers_)), true);
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: fix std::move location

File: source/common/http/http1/codec_impl.cc [hunk 2]
```diff
@@ -626,9 +622,9 @@ void ConnectionImpl::onMessageBeginBase() {
   // protocol for each request. Envoy defaults to 1.1 but sets the protocol to 1.0 where applicable
   // in onHeadersCompleteBase
   protocol_ = Protocol::Http11;
-  ASSERT(!current_header_map_);
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: think it's worth keeping this assert in maybeAlloc...?
I like the guarantees we had before that we never leaked headers/trailers across sequential streams, at least until we move this to ActiveRequest/PendingResponse per your TODO

File: source/common/http/http1/codec_impl.h [hunk 3]
```diff
@@ -213,19 +209,22 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
   Network::Connection& connection_;
   CodecStats stats_;
   http_parser parser_;
-  HeaderMapPtr deferred_end_stream_headers_;
   Http::Code error_code_{Http::Code::BadRequest};
   const HeaderKeyFormatterPtr header_key_formatter_;
   bool processing_trailers_ : 1;
   bool handling_upgrade_ : 1;
   bool reset_stream_called_ : 1;
+  bool deferred_end_stream_headers_ : 1;
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.h]: can we comment this while you're in here?

File: source/common/http/http1/codec_impl.h [hunk 4]
```diff
@@ -363,17 +361,42 @@ class ServerConnectionImpl : public ServerConnection, public ConnectionImpl {
...
+      }
+    }
+  }
 
   ServerConnectionCallbacks& callbacks_;
   std::unique_ptr<ActiveRequest> active_request_;
   Http1Settings codec_settings_;
+  // TODO(mattklein123): This should be a member of ActiveRequest but this change needs dedicated
+  // thought as some of the reset and no header code paths make this difficult.
+  absl::variant<RequestHeaderMapImplPtr, RequestTrailerMapImplPtr> headers_or_trailers_;
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.h]: Comment on lifetime of it having headers/trailers present?

**Comment by mattklein123** [comment]
I'm running into some issues around TestHeaderMap in my next change and I want to make sure I sort that out before I finalize this PR. I will update when I sort that out.

**Comment by mattklein123** [comment]
@alyssawilk updated per comment!

File: source/common/http/http1/codec_impl.cc [hunk 5]
```diff
@@ -784,6 +787,9 @@ void ServerConnectionImpl::onMessageComplete() {
       Buffer::OwnedImpl buffer;
       active_request_->request_decoder_->decodeData(buffer, true);
     }
+
+    // Reset for assertion purposes.
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: IMO reset to ensure no information from one requests persists to the next :-)
Optional, but if you're up for updating, you can do in one of your follow-ups too.

Comment (mattklein123) [reply source/common/http/http1/codec_impl.cc]: Sure I will roll this into my next change.

**Pull request merged**